File: Hard/3919-network-recovery-pathways/network-recovery-pathways.java
class Solution{
    boolean solve(List<List<long[]>> adj, long mid, long k, int n){
        long[] dist=new long[n];
        Arrays.fill(dist, k+1);

        PriorityQueue<long[]> pq=new PriorityQueue<>((a, b)->Long.compare(a[0], b[0]));
        dist[0]=0;
        pq.offer(new long[]{0, 0});

        while(!pq.isEmpty()){
            long[] curr=pq.poll();
            
            long cost=curr[0];
            int node=(int)curr[1];

            if(cost>dist[node])
                continue;

            if(node == n-1)
                return true;

            for(long[] it: adj.get(node)){
                int next=(int)it[0];
                long wt=it[1];

                if(wt<mid)
                    continue;
                if(cost+wt<dist[next]){
                    dist[next]=wt+cost;
                    pq.offer(new long[]{dist[next], next});
                }
            }

        }

        return false;


    }

    public int findMaxPathScore(int[][] e, boolean[] online, long k){
        int n=online.length;

        List<List<long[]>> adj=new ArrayList<>();
        for(int i=0; i<n; i++){
            adj.add(new ArrayList<>());
        }
        
        long low=0, high=0;
        for(int[] it: e){
            high=Math.max(high, it[2]);
            if(online[it[0]] == false || online[it[1]] ==  false)
                continue;
            adj.get(it[0]).add(new long[]{it[1], it[2]});
        }

        long res=-1;
        while(low<=high){
            long mid=low+(high-low)/2;
            if(solve(adj, mid, k, n)){
                res=mid;
                low=mid+1;
            }
            else{
                high=mid-1;
            }
        }

        return (int)res;
    }
}

File: Hard/3919-network-recovery-pathways/network-recovery-pathways.cpp
#define ll long long int

class Solution{
public:
    bool solve(vector<vector<pair<ll, ll>>> &adj, ll mid, ll k, int n){
        vector<ll> dist(n, k+1);
        priority_queue<pair<ll, ll>, vector<pair<ll, ll>>, greater<pair<ll, ll>> > pq;
        dist[0]=0;
        pq.push({0, 0});

        while(!pq.empty()){
            ll cost=pq.top().first;
            ll node=pq.top().second;
            pq.pop();

            if(cost>dist[node])
                continue;

            if(node == n-1)
                return true;

            for(auto it: adj[node]){
                if(it.second<mid)
                    continue;
                if(cost+it.second<dist[it.first]){
                    dist[it.first]=it.second+cost;
                    pq.push({dist[it.first], it.first});
                }
            }

        }

        return false;


    }

    int findMaxPathScore(vector<vector<int>> &e, vector<bool> &online, long long k){
        int n=online.size();

        vector<vector<pair<ll, ll>>> adj(n);
        ll low=0, high=0;
        for(auto it: e){
            high=max(high, (ll)it[2]);
            if(online[it[0]] == false || online[it[1]] ==  false)
                continue;
            adj[it[0]].push_back({it[1], it[2]});
        }

        ll res=-1;
        while(low<=high){
            ll mid=low+(high-low)/2;
            if(solve(adj, mid, k, n)){
                res=mid;
                low=mid+1;
            }
            else{
                high=mid-1;
            }
        }

        return res;

    }
};